test/-ext-/string/test_set_len.rb

```diff
@@ -63,4 +63,22 @@ def test_coderange_after_trunc
     assert_not_predicate str, :ascii_only?
     assert_equal u, str
   end
+
+  def test_valid_encoding_after_resized
+    s = "\0\0".force_encoding(Encoding::UTF_16BE)
+    str = Bug::String.new(s)
+    assert_predicate str, :valid_encoding?
+    str.resize(1)
+    assert_not_predicate str, :valid_encoding?
+    str.resize(2)
+    assert_predicate str, :valid_encoding?
+    str.resize(3)
+    assert_not_predicate str, :valid_encoding?
+
+    s = "\xDB\x00\xDC\x00".force_encoding(Encoding::UTF_16BE)
+    str = Bug::String.new(s)
+    assert_predicate str, :valid_encoding?
+    str.resize(2)
+    assert_not_predicate str, :valid_encoding?
```

Comment:
This is a test case for shrinking surrogate pair that `len % termlen` is not changed but `valid_encoding?` is changed.
This is covered by `slen > len && ENC_CODERANGE(str) != ENC_CODERANGE_7BIT` branch.